base/socket.jl

```diff
@@ -592,81 +592,193 @@ function show(io::IO, err::DNSError)
-            #    invokelatest(cb, IPv6(ntoh(host[1])))
-            #    break
+    data = uv_req_data(req)
+    if data != C_NULL
+        t = unsafe_pointer_to_objref(data)::Task
+        uv_req_set_data(req, C_NULL)
+        if status != 0 || addrinfo == C_NULL
+            schedule(t, UVError("getaddrinfocb", status))
+        else
+            freeaddrinfo = addrinfo
+            addrs = IPAddr[]
+            while addrinfo != C_NULL
+                sockaddr = ccall(:jl_sockaddr_from_addrinfo, Ptr{Void}, (Ptr{Void},), addrinfo)
+                if ccall(:jl_sockaddr_is_ip4, Int32, (Ptr{Void},), sockaddr) == 1
+                    ip4addr = ccall(:jl_sockaddr_host4, UInt32, (Ptr{Void},), sockaddr)
+                    push!(addrs, IPv4(ntoh(ip4addr)))
+                elseif ccall(:jl_sockaddr_is_ip6, Int32, (Ptr{Void},), sockaddr) == 1
+                    ip6addr = Ref{UInt128}()
+                    scope_id = ccall(:jl_sockaddr_host6, UInt32, (Ptr{Void}, Ptr{UInt128}), sockaddr, ip6addr)
+                    push!(addrs, IPv6(ntoh(ip6addr[])))
+                end
+                addrinfo = ccall(:jl_next_from_addrinfo, Ptr{Void}, (Ptr{Void},), addrinfo)
             end
-            addrinfo = ccall(:jl_next_from_addrinfo, Ptr{Void}, (Ptr{Void},), addrinfo)
+            ccall(:uv_freeaddrinfo, Void, (Ptr{Void},), freeaddrinfo)
+            schedule(t, addrs)
         end
-        ccall(:uv_freeaddrinfo, Void, (Ptr{Void},), freeaddrinfo)
+    else
+        # no owner for this req, safe to just free it
+        Libc.free(req)
     end
-    Libc.free(req)
     nothing
 end
 
-function getaddrinfo(cb::Function, host::String)
+"""
+    getalladdrinfo(host::AbstractString) -> Vector{IPAddr}
+
+Gets all of the IP addresses of the `host`.
+Uses the operating system's underlying getaddrinfo implementation, which may do a DNS lookup.
+"""
+function getalladdrinfo(host::String)
     isascii(host) || error("non-ASCII hostname: $host")
-    callback_dict[cb] = cb
-    status = ccall(:jl_getaddrinfo, Int32, (Ptr{Void}, Cstring, Ptr{UInt8}, Any, Ptr{Void}),
-                   eventloop(), host, C_NULL, cb, uv_jl_getaddrinfocb::Ptr{Void})
-    if status == UV_EINVAL
-        throw(ArgumentError("Invalid uv_getaddrinfo() agument"))
-    elseif status in [UV_ENOMEM, UV_ENOBUFS]
-        throw(OutOfMemoryError())
-    elseif status < 0
-        throw(UVError("uv_getaddrinfo returned an unexpected error code", status))
+    req = Libc.malloc(_sizeof_uv_getaddrinfo)
+    uv_req_set_data(req, C_NULL) # in case we get interrupted before arriving at the wait call
+    status = ccall(:jl_getaddrinfo, Int32, (Ptr{Void}, Ptr{Void}, Cstring, Ptr{Void}, Ptr{Void}),
+                   eventloop(), req, host, #=service=#C_NULL, uv_jl_getaddrinfocb::Ptr{Void})
+    if status < 0
+        Libc.free(req)
+        if status == UV_EINVAL
+            throw(ArgumentError("Invalid getaddrinfo argument"))
```

Comment:
probably should update all these error messages to reference `getalladdrinfo` instead of `getaddrinfo` (though I'm not a huge fan of the `getalladdrinfo` name).

Reply:
The error itself still comes from getaddrinfo